data/transactions/logic/eval.go

```diff
@@ -4492,40 +4502,63 @@ func opAppGlobalDel(cx *EvalContext) error {
 // more than 2 or so, and was often called an "index".  But it was not a
 // basics.AssetIndex or basics.ApplicationIndex.
 
-func (cx *EvalContext) appReference(ref uint64, foreign bool) (basics.AppIndex, error) {
+func (cx *EvalContext) appReference(ref uint64, foreign bool) (aid basics.AppIndex, err error) {
 	if cx.version >= directRefEnabledVersion {
-		if ref == 0 || ref == uint64(cx.appID) {
-			return cx.appID, nil
-		}
-		aid := basics.AppIndex(ref)
-		if cx.availableApp(aid) {
-			return aid, nil
-		}
+		return cx.resolveApp(ref)
+	}
 
-		if cx.version < sharedResourcesVersion {
-			// Allow use of indexes, but this comes last so that clear advice can be
-			// given to anyone who cares about semantics in the first few rounds of
-			// a new network - don't use indexes for references, use the App ID
-			if ref <= uint64(len(cx.txn.Txn.ForeignApps)) {
-				return basics.AppIndex(cx.txn.Txn.ForeignApps[ref-1]), nil
+	// resolveApp is already similarly protected (and must be, since it is
+	// called independently)
+	if cx.Proto.AppForbidLowResources {
+		defer func() {
+			if aid <= lastForbiddenResource && err == nil {
+				err = fmt.Errorf("low App lookup %d", aid)
 			}
+		}()
+	}
+	// Old rules, pre directRefEnabledVersion, when a ref has to be a slot for
+	// some opcodes, and had to be an ID for others.
+	if ref == 0 { // Even back when expected to be a real ID, ref = 0 was current app
+		return cx.appID, nil
+	}
+	if foreign {
+		// In old versions, a foreign reference must be an index in ForeignApps or 0
+		if ref <= uint64(len(cx.txn.Txn.ForeignApps)) {
+			return basics.AppIndex(cx.txn.Txn.ForeignApps[ref-1]), nil
 		}
-	} else {
-		// Old rules
-		if ref == 0 { // Even back when expected to be a real ID, ref = 0 was current app
-			return cx.appID, nil
-		}
-		if foreign {
-			// In old versions, a foreign reference must be an index in ForeignAssets or 0
-			if ref <= uint64(len(cx.txn.Txn.ForeignApps)) {
-				return basics.AppIndex(cx.txn.Txn.ForeignApps[ref-1]), nil
+		return 0, fmt.Errorf("App index %d beyond txn.ForeignApps", ref)
+	}
+	// Otherwise it's direct
+	return basics.AppIndex(ref), nil
+}
+
+// resolveApp figures out what App an integer is referring to, considering 0 as
+// current app first, then uses the integer as is if it is an availableApp, then
+// tries to perform a slot lookup.
+func (cx *EvalContext) resolveApp(ref uint64) (aid basics.AppIndex, err error) {
```

Comment:
For new opcodes we might develop in the future, do you imagine they should use the `resolveApp/Asset` functions, or the `app/assetReference` functions? They both should do the same thing, since future opcodes would be introduced after `directRefEnabledVersion`.

I guess this is my way of saying the naming scheme is a bit confusing.

Reply:
yea, I up the vote for confusion from `resolveApp/appReference`, for they are subtly different, yet one depends on another.  But thankfully I think we will get away from it in future opcode, for explicit IDs will be used?

Reply:
Resolving because we decided explicit IDs should be used for future opcodes, sidestepping the need to choose carefully among the existing methods